### PR TITLE
Upgrade to sound null-safety

### PR DESCRIPTION
**Pull request by kengu**
Moves @escamoteur [nullsafety-branch](https://github.com/escamoteur/flutter_map/tree/nullsafety) in PR #851 so that the community can contribute to it without blocking progress.

### CODE REVIEW

File: lib/src/map/map.dart
Comment (josxha) [inline lib/src/map/map.dart:354]: Is there a reason getBounds() returns a `LatLngBounds?` instead of a `LatLngBounds` object? As far as I can see, null cannot be returned here because `_calculateBounds()` already returns a `LatLngBounds` object.

Comment (escamoteur) [reply lib/src/map/map.dart:354]: Agreed, don't know why I out it there. Same with the `bounds` property but I wonder why do we have a get function and a property?

Comment (josxha) [reply lib/src/map/map.dart:354]: Right.  The same goes for line 164:
```dart
LatLng get center => getCenter() ?? options.center;
```
getCenter() can also never return null. 

It seems that the get properties are generally used and the getBounds() and getCenter() functions are more internal functions. So it's really worth considering if they really make sense.

Comment (kengu) [reply lib/src/map/map.dart:354]: > Agreed, don't know why I out it there. Same with the `bounds` property but I wonder why do we have a get function and a property?

Changing it from function to property would be a breaking. I guess this is a good time to to some breaking improvements as any.

Comment (kengu) [reply lib/src/map/map.dart:354]: @josxha The dart analyzer does not allow `LatLongBounds` as return-type because `_lastBounds` is defined as nullable with `LatLngBounds? _lastBounds`. I don't know if this is an error on the dart-side since the if-check above clearly removes the possibility of returning null. I guess there is something else going on here (like me not understanding every aspect of sound null-safety and thus how to handle cases like these) 

Screenshot 2021-04-25 at 14 18 49

Comment (Marthijs-Berfelo) [reply lib/src/map/map.dart:354]: @kengu regarding the null check, despite the not null check you need to "Cast away the null". So the

Comment (kengu) [reply lib/src/map/map.dart:354]: > @kengu regarding the null check, despite the not null check you need to "Cast away the null". So the
> `return _lastBounds;` should be `return _lastBounds!;`
> 
> also see: https://stackoverflow.com/questions/63253015/what-does-the-exclamation-mark-mean-before-a-function-call
> the official null safety documentation does suggest it's not necessary, however I've encountered similar issues and resolved them similarly.

Check - will look into it

Comment (kengu) [reply lib/src/map/map.dart:354]: Seems reasonable enough to use `!`, event though static safety is lost. See [null assertion operator](https://dart.dev/null-safety/understanding-null-safety#null-assertion-operator)

> Of course, like any cast, using ! comes with a loss of static safety. The cast must be checked at runtime to preserve soundness and it may fail and throw an exception. But you have control over where these casts are inserted, and 
 you can always see them by looking through your code.

I'll change `getCenter()` and `getBounds()`it soon

Comment (kengu) [reply lib/src/map/map.dart]: Changes pushed now

**Comment by kengu** [comment]
> I created the pull request #894 to make flutter_map independent of [flutter_image](https://pub.dev/packages/flutter_image) and use the http package instead. After that, as far as I know, this package would be ready for a release with sound null safety. Since http is probably already integrated in most apps, it would also be a nice little dependency clean up.

Thank you @josxha - I think this is a reasonable way forward, see my [rfc described here](https://github.com/fleaflet/flutter_map/pull/894#pullrequestreview-655533983)

**Comment by kengu** [comment]
@johnpryan We are ready to go!

**Comment by ThexXTURBOXx** [comment]
@kengu Maybe we should wait for @josxha to merge [my PR](https://github.com/josxha/flutter_map/pull/1) into his branch. After that you can merge his (and my) changes, such that these are in this PR as well then?

File: example/lib/pages/network_tile_provider.dart
Comment (ThexXTURBOXx) [inline example/lib/pages/network_tile_provider.dart:59]: We should remove this line, right? This doesn't seem to be true by documentation, right? If that's true, we should rename the provider also.

I think I missed part of the discussion in #894, so just ignore that comment if what I'm saying is wrong....

**Review by ThexXTURBOXx — CHANGES_REQUESTED** [review]

**Comment by ThexXTURBOXx** [comment]
This is everything else I could find. Just some missing TODOs and one last thing from the last discussion :)

Of course, this is not my project or anything, but maybe this helps sorting out some last things before @johnpryan can give his final review

**Comment by JaffaKetchup** [comment]
@kengu also asked me to change the base (merge to) branch on my PR for offline stuff (documentation only) (https://github.com/fleaflet/flutter_map/pull/913) to this branch, so can that be merged as well please :)!
Looking forward to the release!

**Comment by aytunch** [comment]
@johnpryan Hi, do you have a time frame to review this PR? We have a very long pubspec.yaml in our project and only `flutter_map` is not converted to null safety yet. And it is blocking us for updating our other packages. Reading the comments, I see there are others in the same situation too. Would appreciate it a lot if you could review this, thanks.

**Comment by johnpryan** [comment]
Yes, I can take a look today.

File: example/lib/pages/network_tile_provider.dart
Comment (andreandersson) [reply example/lib/pages/network_tile_provider.dart:59]: Based on the comments in the code here: https://github.com/fleaflet/flutter_map/blob/issues/829-nullsafety/lib/src/layer/tile_layer.dart#L116

I would say you are correct, and that the opposite is true - it's not cached. Adding the word "not" in the example would probably be enough?

**Comment by andreandersson** [comment]
I've resolved the comments from @ThexXTURBOXx in #919

**Comment by JaffaKetchup** [comment]
Hi,
Glad and excited to see we're almost there!
Can we also [merge this](https://github.com/fleaflet/flutter_map/pull/913) as part of this release?
Thanks for the great work you guys have done :)

File: example/lib/pages/network_tile_provider.dart
Comment (ThexXTURBOXx) [reply example/lib/pages/network_tile_provider.dart:59]: @andreandersson Is the `CachedNetworkTileProvider` still a thing, i.e. does it still exist? If not, we should also remove that comment in the code. Thanks for spotting that! :)

**Comment by danielocampo2** [comment]
Thanks for all the work you're doing, is there any update on this review? Eagerly waiting for this and the other flutter_map libs to be updated.

Thanks!

**Comment by JaffaKetchup** [comment]
Hi,
Just to let you guys know, I've made some more changes to the caching section that I added in my PR #913. It might now be quite complicated, however the naming of some tile providers have made it difficult, besides the fact that Flutter's own documentation contradicts itself!
I have also slightly renewed the old offline maps section to let the user know that the guide will be using TileMill which is difficult to install and has some question marks hanging over it's repository.

File: lib/src/layer/group_layer.dart
Comment (johnpryan) [reply lib/src/layer/group_layer.dart:10]: merged, thanks

**Comment by JaffaKetchup** [comment]
Um, my #913 PR was just overwritten.

**Comment by johnpryan** [comment]
@JaffaKetchup There were merge conflicts, so I'm re-adding that manually.

**Comment by JaffaKetchup** [comment]
Ok, I've linked the exact changes I made above (click Raw to copy paste) :)

**Comment by andreandersson** [comment]
Same goes with #919 , not sure if all those have been solved yey

**Comment by JaffaKetchup** [comment]
Hooray!
Thanks for the work everyone!

**Comment by aytunch** [comment]
Wow :) 
I am so happy. When will it be available on pub.dev?
Thanks to everyone putting in some great work.

**Comment by JaffaKetchup** [comment]
Yeah same, great if this could be published to pub.dev sooner rather than later.

**Comment by johnpryan** [comment]
Published, thanks everyone!

**Pull request merged**
